File: packages-logicn/logicn-devtools-benchmarks/benchmarks/six-digit-guess/bench.cpp
// six-digit-guess — C++ native
// Compile: g++ -O2 -o bench-guess bench.cpp
#include <cstdint>
#include <chrono>
#include <iostream>
#include <string>
#include <cstring>

static const int CODE_LEN=6;

static void bulls_and_cows(const char* cand, const char* tgt, int& bulls, int& cows) {
    bulls=0; int ca[10]={},ta[10]={};
    for(int i=0;i<CODE_LEN;++i) {
        if(cand[i]==tgt[i]) bulls++;
        else { ca[cand[i]-'0']++; ta[tgt[i]-'0']++; }
    }
    cows=0; for(int d=0;d<10;++d) cows+=std::min(ca[d],ta[d]);
}

int main(int argc,char* argv[]) {
    std::string target="042069"; int max_attempts=2000000;
    for(int i=1;i<argc-1;++i) {
        if(!std::strcmp(argv[i],"--target")) target=argv[i+1];
        if(!std::strcmp(argv[i],"--max"))    max_attempts=std::stoi(argv[i+1]);
    }
    using clk=std::chrono::steady_clock; auto t0=clk::now();
    char code[7]; int attempt=0; bool found=false;
    long long tb=0,tc=0;
    while(attempt<max_attempts) {
        int n=attempt%1000000;
        code[0]='0'+(n/100000)%10; code[1]='0'+(n/10000)%10;
        code[2]='0'+(n/1000)%10;  code[3]='0'+(n/100)%10;
        code[4]='0'+(n/10)%10;    code[5]='0'+n%10; code[6]=0;
        attempt++;
        int b=0,c=0; bulls_and_cows(code,target.c_str(),b,c);
        tb+=b; tc+=c;
        if(b==CODE_LEN){found=true;break;}
    }
    double elapsed=std::chrono::duration<double,std::milli>(clk::now()-t0).count();
    std::cout<<"{"
        <<"\"runtime\":\"cpp\","
        <<"\"benchmark\":\"six-digit-guess-v2\","
        <<"\"found\":"<<(found?"true":"false")<<","
        <<"\"attempts\":"<<attempt<<","
        <<"\"totalBulls\":"<<tb<<","
        <<"\"totalCows\":"<<tc<<","
        <<"\"elapsedMs\":"<<elapsed<<","
        <<"\"attemptsPerSecond\":"<<(attempt/(elapsed/1000.0))
        <<"}"<<std::endl;
}

File: packages-logicn/logicn-devtools-benchmarks/benchmarks/six-digit-guess/bench.rs
// six-digit-guess — Rust native
// Compile: rustc -O -o bench-guess bench.rs
use std::time::Instant;
use std::env;
use std::convert::TryInto;
const CODE_LEN:usize=6;
fn bulls_and_cows(cand:&[u8;CODE_LEN],tgt:&[u8;CODE_LEN])->(i32,i32) {
    let mut bulls=0i32; let mut ca=[0i32;10]; let mut ta=[0i32;10];
    for i in 0..CODE_LEN {
        if cand[i]==tgt[i]{bulls+=1;}
        else{ca[(cand[i]-b'0')as usize]+=1;ta[(tgt[i]-b'0')as usize]+=1;}
    }
    let cows:i32=(0..10).map(|d|ca[d].min(ta[d])).sum();
    (bulls,cows)
}
fn fmt_code(n:usize)->[u8;CODE_LEN]{
    [b'0'+((n/100000)%10)as u8,b'0'+((n/10000)%10)as u8,
     b'0'+((n/1000)%10)as u8,b'0'+((n/100)%10)as u8,
     b'0'+((n/10)%10)as u8,b'0'+(n%10)as u8]
}
fn main(){
    let args:Vec<String>=env::args().collect();
    let mut ts="042069".to_string(); let mut max=2_000_000usize;
    let mut i=1;
    while i+1<args.len(){match args[i].as_str(){"--target"=>ts=args[i+1].clone(),"--max"=>max=args[i+1].parse().unwrap_or(max),_=>{}};i+=2;}
    let tgt:[u8;CODE_LEN]=ts.as_bytes()[..CODE_LEN].try_into().unwrap();
    let t0=Instant::now(); let mut attempt=0usize; let mut found=false;
    let mut tb=0i64; let mut tc=0i64;
    while attempt<max {
        let code=fmt_code(attempt%1_000_000); attempt+=1;
        let(b,c)=bulls_and_cows(&code,&tgt);tb+=b as i64;tc+=c as i64;
        if b as usize==CODE_LEN{found=true;break;}
    }
    let elapsed=t0.elapsed().as_secs_f64()*1000.0;
    println!("{{\"runtime\":\"rust\",\"benchmark\":\"six-digit-guess-v2\",\"found\":{},\"attempts\":{},\"totalBulls\":{},\"totalCows\":{},\"elapsedMs\":{:.3},\"attemptsPerSecond\":{:.2}}}",
        found,attempt,tb,tc,elapsed,attempt as f64/(elapsed/1000.0));
}
